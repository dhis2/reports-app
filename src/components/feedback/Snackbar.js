import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { FeedbackSnackbar } from '@dhis2/d2-ui-core'
import { clearFeedback } from '../../redux/actions/feedback'

const Snackbar = ({
    showSnackbar,
    message,
    type,
    action,
    onActionClick,
    onClose,
}) => {
    const conf = { message, type, action, onActionClick }
    return (
        <span id="feedbackSnackbarId">
            <FeedbackSnackbar
                onClose={onClose}
                show={showSnackbar}
                conf={conf}
            />
        </span>
    )
}

Snackbar.propTypes = {
    message: PropTypes.string.isRequired,
    showSnackbar: PropTypes.bool.isRequired,
    type: PropTypes.string.isRequired,
    onClose: PropTypes.func.isRequired,
    action: PropTypes.string,
    onActionClick: PropTypes.func,
}

Snackbar.defaultProps = {
    action: '',
    onActionClick: () => null,
}

const mapStateToProps = ({ feedback }) => ({
    showSnackbar: feedback.showSnackbar,
    message: feedback.message,
    type: feedback.type,
})

const mapDispatchToProps = {
    onClose: clearFeedback,
}

const connectedSnackbar = connect(mapStateToProps, mapDispatchToProps)(Snackbar)

export { connectedSnackbar as Snackbar, Snackbar as SnackbarOriginal }
