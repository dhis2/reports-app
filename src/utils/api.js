import { isDevelopment } from './env/isDevelopment'
import {
    addFileResourceUrlToResource,
    uploadFile,
    standardReportsFields,
    addFilterForName,
    formatStandardReportsResponse,
    mapCollectionToDimensionQueryString,
    mapResponseToArrayOfIds,
    getAnalyticsFileUrls,
    buildQueryString,
    getFileUrls,
    postDataSetReportCommentUrl,
    getDataSetReportFileUrls,
} from './api/helpers'
import { isCustomFormType } from './dataSetReport/isCustomFormType'

export const RESOURCE_ENDPOINT = 'documents'
const DATA_DIMENSION_SUFFIXES = [
    'ACTUAL_REPORTS',
    'EXPECTED_REPORTS',
    'REPORTING_RATE',
    'ACTUAL_REPORTS_ON_TIME',
    'REPORTING_RATE_ON_TIME',
]
const STANDARD_REPORTS_ENDPOINT = 'reports'

let d2
let api

/**
 * Sets d2 and the api
 */
export const initApi = d2Instance => {
    d2 = d2Instance
    api = d2.Api.getApi()

    if (isDevelopment()) {
        window.d2 = d2
        window.d2Api = api
    }
}

/**
 * @return {Object} d2 instance
 */
export const getD2 = () => d2

/**
 * @return {Object} d2 api instance
 */
export const getApi = () => api

/**
 * @returns {string} context path
 */
export const getContextPath = () => d2.system.systemInfo.contextPath

/**
 * @typedef {Object} Version
 * @property {number} major The major version
 * @property {number} minor The minor version
 * @property {boolean} snapshot Flag to indicate if build is a snapshot
 */

/**
 * @returns {Version} The dhis2-core instance version
 */
export const getSystemVersion = () => d2.system.version

/**
 * @return {Promise} Period types
 */
export const getPeriodTypes = () =>
    api.get('periodTypes').then(resp => resp.periodTypes)

/**
 * @return {Promise} Organisation units
 */
export const getOrganisationUnits = () =>
    d2.models.organisationUnits
        .list({
            paging: false,
            level: 1,
            fields: [
                'id',
                'displayName',
                'path',
                'children::isNotEmpty',
                'memberCount',
            ],
        })
        .then(modelCollection => modelCollection.toArray())

/**
 * @param {Array} dataSetOptions
 * @param {Array} orgUnitGroupsOptions
 * @param {string} dataSetId
 * @param {string} orgUnit
 * @param {string} period
 * @param {bool} selectedUnitOnly
 * @returns {Promise}
 */
export const getDataSetReport = options => {
    const endPoint = 'dataSetReport'
    const url = isCustomFormType(options.dataSet.formType)
        ? `${endPoint}/custom`
        : endPoint
    const data = {
        ds: options.dataSet.id,
        pe: options.period,
        ou: options.orgUnit,
        selectedUnitOnly: options.selectedUnitOnly,
        filter: mapCollectionToDimensionQueryString(
            options.dataSetDimensions,
            options.orgUnitGroupsOptions
        ),
    }
    const requestOptions = isCustomFormType(options.dataSet.formType)
        ? { headers: { Accept: 'text/html' } }
        : {}
    const fileUrls = getDataSetReportFileUrls(endPoint, options)

    return api
        .get(url, data, requestOptions)
        .then(response => ({ data: response, fileUrls }))
}

/**
 * @param {string} dataSetId
 * @returns {Promise<array>} The dimensions for a given dataset
 */
export const getDimensions = dataSetId =>
    api
        .get(`dimensions/dataSet/${dataSetId}`, {
            fields: ['id', 'displayName', 'items[id,displayName]'].join(','),
            order: 'name:asc',
            paging: false,
        })
        .then(({ dimensions }) => dimensions)

/**
 * @param {string} dataSetId
 * @param {string} orgUnitId
 * @param {string} period
 * @param {string} comment
 * @returns {Promise}
 */
export const postDataSetReportComment = (
    dataSetId,
    orgUnitId,
    period,
    comment
) => {
    const endpointUrl = postDataSetReportCommentUrl(
        dataSetId,
        orgUnitId,
        period
    )
    const requestHeaders = { headers: { 'Content-Type': 'text/plain' } }
    return api.post(endpointUrl, comment, requestHeaders)
}

/**
 * @param {Object} orgUnit
 * @param {string} dataSetId
 * @param {string} period
 * @param {Object} orgUnitOptions
 * @returns {Promise}
 */
export const getReportingRateSummaryReport = async (
    orgUnit,
    dataSetId,
    period,
    orgUnitOptions
) => {
    const orgUnitIds = await getOrgUnitAndChildrenIds(orgUnit)
    const dataDimensions = DATA_DIMENSION_SUFFIXES.map(
        suffix => `${dataSetId}.${suffix}`
    )
    const req = new d2.analytics.request()
        .addDataDimension(dataDimensions)
        .addOrgUnitDimension(orgUnitIds)
        .addPeriodFilter(period)
        .withColumns('dx')
        .withRows('ou')
        .withTableLayout()
        .withHideEmptyRows()
        .withDisplayProperty('SHORTNAME')
        .withIncludeNumDen(false)

    for (let key in orgUnitOptions) {
        if (orgUnitOptions[key]) {
            req.addFilter(key, orgUnitOptions[key])
        }
    }

    // Instead of calling `d2.analytics.aggregate.get(req)`, which spawns two parallel requests,
    // we're just building the .json url from the request instance and call the regular `api.get(url)`,
    // which only spawns a single request
    const extensions = ['json', 'xls', 'csv']
    const [{ url }, ...fileUrls] = getAnalyticsFileUrls(req, extensions)
    return api.get(url).then(data => ({ ...data, fileUrls }))
}

/**
 * @param {Object} orgUnit
 * @returns {Promise} - Array of IDs of the orgUnit and its direct descendants
 */
export const getOrgUnitAndChildrenIds = orgUnit => {
    const children = orgUnit.children.size
        ? Promise.resolve(orgUnit.children)
        : d2.models.organisationUnits
              .get(orgUnit.id, { fields: ['children[id]'] })
              .then(({ children }) => children)

    return children.then(children => [
        orgUnit.id,
        ...mapResponseToArrayOfIds(children),
    ])
}

/**
 * @returns {Promise}
 */
export const getOrgUnitGroupSets = () =>
    d2.models.organisationUnitGroupSet.list({
        paging: false,
        fields: 'id,displayName,organisationUnitGroups[id,displayName]',
    })

/**
 * @param {Object} orgUnit
 * @param {string} groupSetId
 * @returns {Promise}
 */
export const getOrgUnitDistReport = async (
    orgUnit,
    groupSetId,
    shouldShowChart
) => {
    const orgUnitIds = shouldShowChart
        ? orgUnit.id
        : await getOrgUnitAndChildrenIds(orgUnit)

    const endPoint = 'orgUnitAnalytics'
    const queryString = buildQueryString({
        ou: orgUnitIds,
        ougs: groupSetId,
        columns: groupSetId,
    })
    const relativeUrl = `${endPoint}?${queryString}`
    const fileUrls = getFileUrls(endPoint, queryString, ['xls', 'pdf'])

    return api.get(relativeUrl).then(response => ({ ...response, fileUrls }))
}

/**
 * @returns {Promise}
 */
export const getResources = (page, pageSize, search) => {
    const requestData = {
        page,
        pageSize,
        fields: 'displayName,id,url,external,access',
    }

    if (search) {
        requestData.filter = `displayName:ilike:${search}`
    }

    return api.get(RESOURCE_ENDPOINT, requestData)
}

/**
 * @returns {Promise}
 */
export const deleteResource = resourceId =>
    api.delete(`${RESOURCE_ENDPOINT}/${resourceId}`)

/**
 * @returns {Promise}
 */
export const getStandardReportTables = () =>
    api.get('reportTables', { paging: false, fields: 'id,name' })

/**
 * returns {Promise}
 */
export const getStandardReportTable = (id, queryParams = {}) =>
    api.get(`reportTables/${id}`, queryParams)

/**
 * @param {Object} report
 * @returns {Promise}
 */
export const postStandardReport = report =>
    api.post(STANDARD_REPORTS_ENDPOINT, report)

/**
 * @param {Object} report
 * @returns {Promise}
 */
export const updateStandardReport = report =>
    api.update(`${STANDARD_REPORTS_ENDPOINT}/${report.id}`, report)

/**
 * @returns {Promise}
 */
export const getDataSetOptions = () =>
    d2.models.dataSet
        .list({ paging: false, fields: 'id,displayName,formType' })
        .then(response => response.toArray())

/**
 * @param {Object} api
 * @param {Object} resource
 * @returns {Promise}
 */
export const postDocument = (api, resource) =>
    api.post(RESOURCE_ENDPOINT, resource)

/**
 * @param {Object} api
 * @param {string} resourceId
 * @param {Object} resource
 * @returns {Promise}
 */
export const putDocument = (api, resourceId, resource) =>
    api.update(`${RESOURCE_ENDPOINT}/${resourceId}`, resource)

/**
 * @param {Object} resource
 * @param {File} [file] - an optional file
 * @returns {Promise}
 */
export const postResource = (resource, file = null) =>
    file
        ? uploadFile(api, file)
              .then(({ response }) =>
                  Promise.resolve(
                      addFileResourceUrlToResource(
                          resource,
                          response.fileResource
                      )
                  )
              )
              .then(resource => postDocument(api, resource))
        : postDocument(api, resource)

/**
 * @param {Object} resource
 * @param {File} [file] - an optional file
 * @returns {Promise}
 */
export const putResource = (resourceId, resource, file = null) =>
    file
        ? uploadFile(api, file)
              .then(({ response }) =>
                  Promise.resolve(
                      addFileResourceUrlToResource(
                          resource,
                          response.fileResource
                      )
                  )
              )
              .then(resource => putDocument(api, resourceId, resource))
        : putDocument(api, resourceId, resource)

/**
 * =================================
 * Standard report
 * =================================
 */

/**
 * @param {number} page
 * @param {number} pageSize
 * @param {string} nameFilter
 * @return {Promise}
 */
export const getFilteredStandardReports = (page, pageSize, nameFilter) =>
    addFilterForName(nameFilter, d2.models.report)
        .list({ page, pageSize, fields: standardReportsFields })
        .then(formatStandardReportsResponse)

/**
 * @param {string} id
 * @returns {Promise}
 */
export const getStandardReportDetails = id =>
    api.get(`${STANDARD_REPORTS_ENDPOINT}/${id}`)

/**
 * @param {string} id
 * @return {Promise}
 */
export const deleteStandardReport = id =>
    api.delete(`${STANDARD_REPORTS_ENDPOINT}/${id}`)

/**
 * @param {string} id
 * @returns {Promise}
 */
export const getStandardReportHtmlReport = (id, requestBody) =>
    api.get(`reports/${id}/data.html`, {
        t: new Date().getTime(),
        ...requestBody,
    })
