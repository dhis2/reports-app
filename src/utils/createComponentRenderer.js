import React from 'react'
import { shallow } from 'enzyme'

export default function createComponentRenderer(Component, defaultProps) {
    return function(customProps) {
        const props = { ...defaultProps, ...customProps }
        return shallow(<Component {...props} />)
    }
}
