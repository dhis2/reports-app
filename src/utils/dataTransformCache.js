export default function createDataTransformCache() {
    let inputData = null
    let transformedData = null
    return {
        hasValidCacheFor: data => !!inputData && data === inputData,
        getCachedResult: () => transformedData,
        setCachedResult: (input, transformed) => {
            inputData = input
            transformedData = transformed
        },
    }
}
