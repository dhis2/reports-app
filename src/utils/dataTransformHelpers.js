/**
 * @param {Object} state - app state
 * @returns {String} - table title
 */
export function getTitle(state) {
    return [
        state.reportData.content.headers[1].column,
        state.organisationUnits.selected.displayName,
    ].join(' - ')
}

/**
 * @param {Object} data - Response data from the server
 * @returns {Object} - Containing rows and columns array, columns sorted ASC
 */
export function createGrid(data) {
    // As new Set for constant time look-up
    const orgUnitLookup = new Set(data.orgUnitIds)
    const items = data.metaData.items
    const { rows, columns } = Object.keys(items).reduce(
        (acc, id) => {
            const targetArray = orgUnitLookup.has(id) ? acc.rows : acc.columns
            targetArray.push({ id: id, name: items[id].name })
            return acc
        },
        { rows: [], columns: [] }
    )
    return {
        rows,
        columns: columns.sort((a, b) =>
            // Same sorting as struts app
            a.name.toUpperCase().localeCompare(b.name.toUpperCase())
        ),
    }
}
