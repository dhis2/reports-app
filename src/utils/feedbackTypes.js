export const ACTION_MESSAGE = 'ACTION_MESSAGE'
// TODO: remove loading
export const LOADING = 'LOADING'
export const ERROR = 'ERROR'
export const SUCCESS = 'SUCCESS'
export const WARNING = 'WARNING'
