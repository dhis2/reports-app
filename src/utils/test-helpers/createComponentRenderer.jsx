import { shallow } from 'enzyme'
import React from 'react'

export default function createComponentRenderer(Component, defaultProps) {
    return function (customProps) {
        const props = { ...defaultProps, ...customProps }
        return shallow(<Component {...props} />)
    }
}
