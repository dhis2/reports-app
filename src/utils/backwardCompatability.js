import { getSystemMinorVersion } from './api'

/*
The following API changes were introduced in v2.34:

1. The `api/reportTables` endpoint was replaced by `api/visualizations`

2. The response payload from this endpoint now looks slightly different:
    
    < 2.34
    "reportParams": {
        "paramGrandParentOrganisationUnit": false,
        "paramReportingPeriod": false,
        "paramOrganisationUnit": false,
        "paramParentOrganisationUnit": false
    }

    >= 2.34
    "reportingParams": {
        "grandParentOrganisationUnit": false,
        "reportingPeriod": false,
        "organisationUnit": false,
        "parentOrganisationUnit": false
    }

3. From >= 2.34 onward we have to deal with some inconsistencies in the property names:
    - `report.reportParams`
    - `report.visualization.reportingParams`

The functions in this module are there to deal with these discrepancies. To avoid too many changes to the app state model, the following names have been settled on:
1. `reportTables` in favour of `visualizations`
2. `reportParams` in favour of `reportingParams`
3. don't prefix reportParams properties with `param`, 
   i.e. `reportingPeriod` in favour of `paramReportingPeriod`


When we drop support for version v2.33 these helpers can be removed and replaced
with much more simple and sane logic
*/

const isBelowVersion34 = () => getSystemMinorVersion() < 34
const isAtLeastVersion34 = () => !isBelowVersion34()

const getReportTableNameBySystemVersion = () =>
    isBelowVersion34() ? 'reportTable' : 'visualization'

const getReportParamsFieldName = () =>
    isBelowVersion34() ? 'reportParams' : 'reportingParams'

export const getReportTablesResourceNameBySystemVersion = () =>
    `${getReportTableNameBySystemVersion()}s`

export const getReportTablesFilterBySystemVersion = searchTerm =>
    isAtLeastVersion34()
        ? [`identifiable:token:${searchTerm}`, 'type:eq:PIVOT_TABLE']
        : [`name:ilike:${searchTerm}`]

/**
 * Required fields for displaying the standard reports
 */
export const getStandardReportsFieldsBySystemVersion = () => {
    const reportTableFieldName = getReportTableNameBySystemVersion()
    const reportParamsFieldName = getReportParamsFieldName()

    return [
        'displayName',
        'type',
        'id',
        `${reportTableFieldName}[id,displayName,${reportParamsFieldName},relativePeriods]`,
        'reportParams',
        'relativePeriods',
        'access',
    ]
}

const getReportParamsPropertiesBySystemVersion = reportParams => {
    if (isAtLeastVersion34() || !reportParams) {
        return reportParams
    }

    return {
        reportingPeriod: reportParams.paramReportingPeriod,
        grandParentOrganisationUnit:
            reportParams.paramGrandParentOrganisationUnit,
        organisationUnit: reportParams.paramOrganisationUnit,
        parentOrganisationUnit: reportParams.paramParentOrganisationUnit,
    }
}

export const getStandardReportFieldsBySystemVersion = () => {
    const reportTableField = isAtLeastVersion34()
        ? 'visualization[id,displayName]'
        : 'reportTable[id,displayName]'

    return [':owner', reportTableField]
}

export const formatStandardReportResponseBySystemVersion = reportModel => {
    // handle both d2 model instances and plain objects
    const reportJson = reportModel.toJSON ? reportModel.toJSON() : reportModel
    // <2.34 reportModel.reportTable, >=2.34 reportmodel.visualization
    const reportTable = reportModel[getReportTableNameBySystemVersion()]
    // <2.34 reportModel.reportTable.reportParams, >=2.34 reportmodel.visualization.reportingParams
    const reportTableReportParams =
        reportTable && reportTable[getReportParamsFieldName()]

    return {
        ...reportJson,
        reportParams: getReportParamsPropertiesBySystemVersion(
            reportJson.reportParams
        ),
        reportTable: {
            // The JSON representation of a d2 reportModel is missing
            // some of the reportTable properties such as reportParams
            ...reportTable,
            reportParams: getReportParamsPropertiesBySystemVersion(
                reportTableReportParams
            ),
        },
    }
}

export const formatStandardReportPayloadBySystemVersion = report => {
    if (isAtLeastVersion34()) {
        const visualization = report.reportTable
        delete report.reportTable
        return {
            ...report,
            visualization,
        }
    }

    const rp = report.reportParams

    return {
        ...report,
        reportParams: {
            paramGrandParentOrganisationUnit: rp.grandParentOrganisationUnit,
            paramOrganisationUnit: rp.organisationUnit,
            paramParentOrganisationUnit: rp.parentOrganisationUnit,
            paramReportingPeriod: rp.reportingPeriod,
        },
    }
}
