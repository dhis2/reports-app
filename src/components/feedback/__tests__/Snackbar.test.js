import createComponentRenderer from '../../../utils/test-helpers/createComponentRenderer'
import { SUCCESS, ERROR, WARNING } from '../../../utils/feedbackTypes'
import { SnackbarOriginal as Snackbar } from '../Snackbar'

jest.mock('@dhis2/d2-ui-core', () => ({
    SnackbarSnackbar: 'SnackbarSnackbar',
    CircularProgress: 'CircularProgress',
}))

describe('<Snackbar />', () => {
    const defaultProps = {
        showSnackbar: false,
        showLoader: false,
        message: '',
        type: '',
        action: undefined,
        onClose: jest.fn(),
    }
    const renderSnackbar = createComponentRenderer(Snackbar, defaultProps)

    // for whatever reason jest can't import the feedbacksnackbar from d2-ui
    // and throws console errors....
    jest.spyOn(console, 'error').mockImplementation(() => null)

    afterEach(() => {
        jest.clearAllMocks()
    })

    it('can render a default snackbar message', () => {
        expect(
            renderSnackbar({ showSnackbar: true, message: 'Message body' })
        ).toMatchSnapshot()
    })

    it('can render a success snackbar message', () => {
        expect(
            renderSnackbar({
                showSnackbar: true,
                message: 'Message body',
                type: SUCCESS,
            })
        ).toMatchSnapshot()
    })

    it('can render a warning snackbar message', () => {
        expect(
            renderSnackbar({
                showSnackbar: true,
                message: 'Message body',
                type: WARNING,
            })
        ).toMatchSnapshot()
    })

    it('can render an error snackbar message', () => {
        expect(
            renderSnackbar({
                showSnackbar: true,
                message: 'Message body',
                type: ERROR,
            })
        ).toMatchSnapshot()
    })
})
