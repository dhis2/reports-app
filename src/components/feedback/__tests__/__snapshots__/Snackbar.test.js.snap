// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Snackbar /> can render a default snackbar message 1`] = `
<span
  id="feedbackSnackbarId"
>
  <Component
    conf={
      Object {
        "action": "",
        "message": "Message body",
        "onActionClick": [Function],
        "type": "",
      }
    }
    onClose={[MockFunction]}
    show={true}
  />
</span>
`;

exports[`<Snackbar /> can render a success snackbar message 1`] = `
<span
  id="feedbackSnackbarId"
>
  <Component
    conf={
      Object {
        "action": "",
        "message": "Message body",
        "onActionClick": [Function],
        "type": "SUCCESS",
      }
    }
    onClose={[MockFunction]}
    show={true}
  />
</span>
`;

exports[`<Snackbar /> can render a warning snackbar message 1`] = `
<span
  id="feedbackSnackbarId"
>
  <Component
    conf={
      Object {
        "action": "",
        "message": "Message body",
        "onActionClick": [Function],
        "type": "WARNING",
      }
    }
    onClose={[MockFunction]}
    show={true}
  />
</span>
`;

exports[`<Snackbar /> can render an error snackbar message 1`] = `
<span
  id="feedbackSnackbarId"
>
  <Component
    conf={
      Object {
        "action": "",
        "message": "Message body",
        "onActionClick": [Function],
        "type": "ERROR",
      }
    }
    onClose={[MockFunction]}
    show={true}
  />
</span>
`;
